archived/mysql-5.1.36/man/mysqld_safe.1
.\"     Title: \fBmysqld_safe\fR
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.70.1 <http://docbook.sf.net/>
.\"      Date: 06/16/2009
.\"    Manual: MySQL Database System
.\"    Source: MySQL 5.1
.\"
.TH "\fBMYSQLD_SAFE\fR" "1" "06/16/2009" "MySQL 5.1" "MySQL Database System"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
mysqld_safe \- MySQL server startup script
.SH "SYNOPSIS"
.HP 20
\fBmysqld_safe \fR\fB\fIoptions\fR\fR
.SH "DESCRIPTION"
.PP
\fBmysqld_safe\fR
is the recommended way to start a
\fBmysqld\fR
server on Unix and NetWare.
\fBmysqld_safe\fR
adds some safety features such as restarting the server when an error occurs and logging runtime information to an error log file. Descriptions of error logging and NetWare\-specific behaviors are given later in this section.
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBNote\fR
.PP
In MySQL 5.1.20 (only), the default error logging behavior with
\fBmysqld_safe\fR
is to write errors to
syslog
on systems that support the
\fBlogger\fR
program. This differs from the default behavior of writing an error log file for other versions.
.PP
\fBIn 5.1.20, logging to \fR\fBsyslog\fR\fB may fail to operate correctly in some cases; if so, use \fR\fB\fB\-\-skip\-syslog\fR\fR\fB to use the default log file or \fR\fB\fB\-\-log\-error=\fR\fB\fIfile_name\fR\fR\fR\fB to specify a log file name explicitly.\fR
.PP
\fBmysqld_safe\fR
tries to start an executable named
\fBmysqld\fR. To override the default behavior and specify explicitly the name of the server you want to run, specify a
\fB\-\-mysqld\fR
or
\fB\-\-mysqld\-version\fR
option to
\fBmysqld_safe\fR. You can also use
\fB\-\-ledir\fR
to indicate the directory where
\fBmysqld_safe\fR
should look for the server.
.PP
Many of the options to
\fBmysqld_safe\fR
are the same as the options to
\fBmysqld\fR. See
Section\ 5.1.2, \(lqServer Command Options\(rq.
.PP
Options unknown to
\fBmysqld_safe\fR
are passed to
\fBmysqld\fR
if they are specified on the command line, but ignored if they are specified in the
[mysqld_safe]
group of an option file. See
Section\ 4.2.3.3, \(lqUsing Option Files\(rq.
.PP
\fBmysqld_safe\fR
reads all options from the
[mysqld],
[server], and
[mysqld_safe]
sections in option files. For example, if you specify a
[mysqld]
section like this,
\fBmysqld_safe\fR
will find and use the
\fB\-\-log\-error\fR
option:
.sp
.RS 3n
.nf
[mysqld]
log\-error=error.log
.fi
.RE
.PP
For backward compatibility,
\fBmysqld_safe\fR
also reads
[safe_mysqld]
sections, although you should rename such sections to
[mysqld_safe]
in MySQL 5.1 installations.
.PP
\fBmysqld_safe\fR
supports the options in the following list. It also reads option files and supports the options for processing them described at
Section\ 4.2.3.3.1, \(lqCommand\-Line Options that Affect Option\-File Handling\(rq.
.TP 3n
\(bu
\fB\-\-help\fR
.sp
Display a help message and exit.
.TP 3n
\(bu
\fB\-\-autoclose\fR
.sp
(NetWare only) On NetWare,
\fBmysqld_safe\fR
provides a screen presence. When you unload (shut down) the
\fBmysqld_safe\fR
NLM, the screen does not by default go away. Instead, it prompts for user input:
.sp
.RS 3n
.nf
*<NLM has terminated; Press any key to close the screen>*
.fi
.RE
If you want NetWare to close the screen automatically instead, use the
\fB\-\-autoclose\fR
option to
\fBmysqld_safe\fR.
.TP 3n
\(bu
\fB\-\-basedir=\fR\fB\fIpath\fR\fR
.sp
The path to the MySQL installation directory.
.TP 3n
\(bu
\fB\-\-core\-file\-size=\fR\fB\fIsize\fR\fR
.sp
The size of the core file that
\fBmysqld\fR
should be able to create. The option value is passed to
\fBulimit \-c\fR.
.TP 3n
\(bu
\fB\-\-datadir=\fR\fB\fIpath\fR\fR
.sp
The path to the data directory.
.TP 3n
\(bu
\fB\-\-defaults\-extra\-file=\fR\fB\fIpath\fR\fR
.sp
The name of an option file to be read in addition to the usual option files. This must be the first option on the command line if it is used. If the file does not exist or is otherwise inaccessible, the server will exit with an error.
.TP 3n
\(bu
\fB\-\-defaults\-file=\fR\fB\fIfile_name\fR\fR
.sp
The name of an option file to be read instead of the usual option files. This must be the first option on the command line if it is used.
.TP 3n
\(bu
\fB\-\-ledir=\fR\fB\fIpath\fR\fR
.sp
If
\fBmysqld_safe\fR
cannot find the server, use this option to indicate the path name to the directory where the server is located.
.TP 3n
\(bu
\fB\-\-log\-error=\fR\fB\fIfile_name\fR\fR
.sp
Write the error log to the given file. See
Section\ 5.2.2, \(lqThe Error Log\(rq.
.TP 3n
\(bu
\fB\-\-mysqld=\fR\fB\fIprog_name\fR\fR
.sp
The name of the server program (in the
ledir
directory) that you want to start. This option is needed if you use the MySQL binary distribution but have the data directory outside of the binary distribution. If
\fBmysqld_safe\fR
cannot find the server, use the
\fB\-\-ledir\fR
option to indicate the path name to the directory where the server is located.
.TP 3n
\(bu
\fB\-\-mysqld\-version=\fR\fB\fIsuffix\fR\fR
.sp
This option is similar to the
\fB\-\-mysqld\fR
option, but you specify only the suffix for the server program name. The basename is assumed to be
\fBmysqld\fR. For example, if you use
\fB\-\-mysqld\-version=debug\fR,
\fBmysqld_safe\fR
starts the
\fBmysqld\-debug\fR
program in the
ledir
directory. If the argument to
\fB\-\-mysqld\-version\fR
is empty,
\fBmysqld_safe\fR
uses
\fBmysqld\fR
in the
ledir
directory.
.TP 3n
\(bu
\fB\-\-nice=\fR\fB\fIpriority\fR\fR
.sp
Use the
nice
program to set the server's scheduling priority to the given value.
.TP 3n
\(bu
\fB\-\-no\-defaults\fR
.sp
Do not read any option files. This must be the first option on the command line if it is used.
.TP 3n
\(bu
\fB\-\-open\-files\-limit=\fR\fB\fIcount\fR\fR
.sp
The number of files that
\fBmysqld\fR
should be able to open. The option value is passed to
\fBulimit \-n\fR. Note that you need to start
\fBmysqld_safe\fR
as
root
for this to work properly!
.TP 3n
\(bu
\fB\-\-pid\-file=\fR\fB\fIfile_name\fR\fR
.sp
The path name of the process ID file.
.TP 3n
\(bu
\fB\-\-port=\fR\fB\fIport_num\fR\fR
.sp
The port number that the server should use when listening for TCP/IP connections. The port number must be 1024 or higher unless the server is started by the
root
system user.
.TP 3n
\(bu
\fB\-\-skip\-kill\-mysqld\fR
.sp
Do not try to kill stray
\fBmysqld\fR
processes at startup. This option works only on Linux.
.TP 3n
\(bu
\fB\-\-socket=\fR\fB\fIpath\fR\fR
.sp
The Unix socket file that the server should use when listening for local connections.
.TP 3n
\(bu
\fB\-\-syslog\fR,
\fB\-\-skip\-syslog\fR
.sp
\fB\-\-syslog\fR
causes error messages to be sent to
syslog
on systems that support the
\fBlogger\fR
program.
\-\-skip\-syslog
suppresses the use of
syslog; messages are written to an error log file. These options were added in MySQL 5.1.20.
.TP 3n
\(bu
\fB\-\-syslog\-tag=\fR\fB\fItag\fR\fR
.sp
For logging to
syslog, messages from
\fBmysqld_safe\fR
and
\fBmysqld\fR
are written with a tag of
mysqld_safe
and
mysqld, respectively. To specify a suffix for the tag, use
\fB\-\-syslog\-tag=\fR\fB\fItag\fR\fR, which modifies the tags to be
mysqld_safe\-\fItag\fR
and
mysqld\-\fItag\fR. This option was added in MySQL 5.1.21.
.TP 3n
\(bu
\fB\-\-timezone=\fR\fB\fItimezone\fR\fR
.sp
Set the
TZ
time zone environment variable to the given option value. Consult your operating system documentation for legal time zone specification formats.
.TP 3n
\(bu
\fB\-\-user={\fR\fB\fIuser_name\fR\fR\fB|\fR\fB\fIuser_id\fR\fR\fB}\fR
.sp
Run the
\fBmysqld\fR
server as the user having the name
\fIuser_name\fR
or the numeric user ID
\fIuser_id\fR. (\(lqUser\(rq
in this context refers to a system login account, not a MySQL user listed in the grant tables.)
.sp
.RE
.PP
If you execute
\fBmysqld_safe\fR
with the
\fB\-\-defaults\-file\fR
or
\fB\-\-defaults\-extra\-file\fR
option to name an option file, the option must be the first one given on the command line or the option file will not be used. For example, this command will not use the named option file:
.sp
.RS 3n
.nf
mysql> \fBmysqld_safe \-\-port=\fR\fB\fIport_num\fR\fR\fB \-\-defaults\-file=\fR\fB\fIfile_name\fR\fR
.fi
.RE
.PP
Instead, use the following command:
.sp
.RS 3n
.nf
mysql> \fBmysqld_safe \-\-defaults\-file=\fR\fB\fIfile_name\fR\fR\fB \-\-port=\fR\fB\fIport_num\fR\fR
.fi
.RE
.PP
The
\fBmysqld_safe\fR
script is written so that it normally can start a server that was installed from either a source or a binary distribution of MySQL, even though these types of distributions typically install the server in slightly different locations. (See
Section\ 2.1.5, \(lqInstallation Layouts\(rq.)
\fBmysqld_safe\fR
expects one of the following conditions to be true:
.TP 3n
\(bu
The server and databases can be found relative to the working directory (the directory from which
\fBmysqld_safe\fR
is invoked). For binary distributions,
\fBmysqld_safe\fR
looks under its working directory for
\fIbin\fR
and
\fIdata\fR
directories. For source distributions, it looks for
\fIlibexec\fR
and
\fIvar\fR
directories. This condition should be met if you execute
\fBmysqld_safe\fR
from your MySQL installation directory (for example,
\fI/usr/local/mysql\fR
for a binary distribution).
.TP 3n
\(bu
If the server and databases cannot be found relative to the working directory,
\fBmysqld_safe\fR
attempts to locate them by absolute path names. Typical locations are
\fI/usr/local/libexec\fR
and
\fI/usr/local/var\fR. The actual locations are determined from the values configured into the distribution at the time it was built. They should be correct if MySQL is installed in the location specified at configuration time.
.sp
.RE
.PP
Because
\fBmysqld_safe\fR
tries to find the server and databases relative to its own working directory, you can install a binary distribution of MySQL anywhere, as long as you run
\fBmysqld_safe\fR
from the MySQL installation directory:
.sp
.RS 3n
.nf
shell> \fBcd \fR\fB\fImysql_installation_directory\fR\fR
shell> \fBbin/mysqld_safe &\fR
.fi
.RE
.PP
If
\fBmysqld_safe\fR
fails, even when invoked from the MySQL installation directory, you can specify the
\fB\-\-ledir\fR
and
\fB\-\-datadir\fR
options to indicate the directories in which the server and databases are located on your system.
.PP
When you use
\fBmysqld_safe\fR
to start
\fBmysqld\fR,
\fBmysqld_safe\fR
arranges for error (and notice) messages from itself and from
\fBmysqld\fR
to go to the same destination.
.PP
As of MySQL 5.1.20, there are several
\fBmysqld_safe\fR
options for controlling the destination of these messages:
.TP 3n
\(bu
\fB\-\-syslog\fR: Write error messages to
syslog
on systems that support the
\fBlogger\fR
program.
.TP 3n
\(bu
\fB\-\-skip\-syslog\fR: Do not write error messages to
syslog. Messages are written to the default error log file (\fI\fIhost_name\fR\fR\fI.err\fR
in the data directory), or to a named file if the
\fB\-\-log\-error\fR
option is given.
.TP 3n
\(bu
\fB\-\-log\-error=\fR\fB\fIfile_name\fR\fR: Write error messages to the named error file.
.sp
.RE
.PP
If none of these options is given, the default is
\fB\-\-skip\-syslog\fR.
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBNote\fR
.PP
In MySQL 5.1.20
\fIonly\fR, the default is
\fB\-\-syslog\fR. This differs from logging behavior for other versions of MySQL, for which the default is to write messages to the default error log file.
.PP
If
\fB\-\-syslog\fR
and
\fB\-\-log\-error\fR
are both given, a warning is issued and
\fB\-\-log\-error\fR
takes precedence.
.PP
When
\fBmysqld_safe\fR
writes a message, notices go to the logging destination (syslog
or the error log file) and
stdout. Errors go to the logging destination and
stderr.
.PP
Before MySQL 5.1.20, error logging is controlled only with the
\fB\-\-log\-error\fR
option. If it is given, messages go to the named error file. Otherwise, messages go to the default error file.
.PP
Normally, you should not edit the
\fBmysqld_safe\fR
script. Instead, configure
\fBmysqld_safe\fR
by using command\-line options or options in the
[mysqld_safe]
section of a
\fImy.cnf\fR
option file. In rare cases, it might be necessary to edit
\fBmysqld_safe\fR
to get it to start the server properly. However, if you do this, your modified version of
\fBmysqld_safe\fR
might be overwritten if you upgrade MySQL in the future, so you should make a copy of your edited version that you can reinstall.
.PP
On NetWare,
\fBmysqld_safe\fR
is a NetWare Loadable Module (NLM) that is ported from the original Unix shell script. It starts the server as follows:
.TP 3n
1.
Runs a number of system and option checks.
.TP 3n
2.
Runs a check on
MyISAM
tables.
.TP 3n
3.
Provides a screen presence for the MySQL server.
.TP 3n
4.
Starts
\fBmysqld\fR, monitors it, and restarts it if it terminates in error.
.TP 3n
5.
Sends error messages from
\fBmysqld\fR
to the
\fI\fIhost_name\fR\fR\fI.err\fR
file in the data directory.
.TP 3n
6.
Sends
\fBmysqld_safe\fR
screen output to the
\fI\fIhost_name\fR\fR\fI.safe\fR
file in the data directory.
.SH "COPYRIGHT"
.PP
Copyright 2007\-2008 MySQL AB, 2009 Sun Microsystems, Inc.
.PP
This documentation is free software; you can redistribute it and/or modify it only under the terms of the GNU General Public License as published by the Free Software Foundation; version 2 of the License.
.PP
This documentation is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along with the program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110\-1301 USA or see http://www.gnu.org/licenses/.
.SH "SEE ALSO"
For more information, please refer to the MySQL Reference Manual,
which may already be installed locally and which is also available
online at http://dev.mysql.com/doc/.
.SH AUTHOR
Sun Microsystems, Inc. (http://www.mysql.com/).


archived/mysql-4.0.19/scripts/mysqld_safe.sh
#!/bin/sh
# Copyright Abandoned 1996 TCX DataKonsult AB & Monty Program KB & Detron HB
# This file is public domain and comes with NO WARRANTY of any kind
#
# scripts to start the MySQL daemon and restart it if it dies unexpectedly
#
# This should be executed in the MySQL base directory if you are using a
# binary installation that has other paths than you are using.
#
# mysql.server works by first doing a cd to the base directory and from there
# executing mysqld_safe

#KILL_MYSQLD=1;
KILL_MYSQLD=0;

trap '' 1 2 3 15			# we shouldn't let anyone kill us

umask 007

defaults=
case "$1" in
    --no-defaults|--defaults-file=*|--defaults-extra-file=*)
      defaults="$1"; shift
      ;;
esac

parse_arguments() {
  # We only need to pass arguments through to the server if we don't
  # handle them here.  So, we collect unrecognized options (passed on
  # the command line) into the args variable.
  pick_args=
  if test "$1" = PICK-ARGS-FROM-ARGV
  then
    pick_args=1
    shift
  fi

  for arg do
    case "$arg" in
      --skip-kill-mysqld*)
        KILL_MYSQLD=0;
        ;;
      # these get passed explicitly to mysqld
      --basedir=*) MY_BASEDIR_VERSION=`echo "$arg" | sed -e "s;--basedir=;;"` ;;
      --datadir=*) DATADIR=`echo "$arg" | sed -e "s;--datadir=;;"` ;;
      --pid-file=*) pid_file=`echo "$arg" | sed -e "s;--pid-file=;;"` ;;
      --user=*)
        if test $SET_USER -eq 0
        then
          user=`echo "$arg" | sed -e "s;--[^=]*=;;"`
        fi
        SET_USER=1
        ;;

      # these two might have been set in a [mysqld_safe] section of my.cnf
      # they are added to mysqld command line to override settings from my.cnf
      --socket=*)  mysql_unix_port=`echo "$arg" | sed -e "s;--socket=;;"` ;;
      --port=*)    mysql_tcp_port=`echo "$arg" | sed -e "s;--port=;;"` ;;

      # mysqld_safe-specific options - must be set in my.cnf ([mysqld_safe])!
      --ledir=*)   ledir=`echo "$arg" | sed -e "s;--ledir=;;"` ;;
      # err-log should be removed in 5.0
      --err-log=*) err_log=`echo "$arg" | sed -e "s;--err-log=;;"` ;;
      --log-error=*) err_log=`echo "$arg" | sed -e "s;--log-error=;;"` ;;
      # QQ The --open-files should be removed in 5.0
      --open-files=*) open_files=`echo "$arg" | sed -e "s;--open-files=;;"` ;;
      --open-files-limit=*) open_files=`echo "$arg" | sed -e "s;--open-files-limit=;;"` ;;
      --core-file-size=*) core_file_size=`echo "$arg" | sed -e "s;--core-file-size=;;"` ;;
      --timezone=*) TZ=`echo "$arg" | sed -e "s;--timezone=;;"` ; export TZ; ;;
      --mysqld=*)   MYSQLD=`echo "$arg" | sed -e "s;--mysqld=;;"` ;;
      --mysqld-version=*)
	tmp=`echo "$arg" | sed -e "s;--mysqld-version=;;"`
	if test -n "$tmp"
	then
	  MYSQLD="mysqld-$tmp"
	else
	  MYSQLD="mysqld"
	fi
	;;
      --nice=*) niceness=`echo "$arg" | sed -e "s;--nice=;;"` ;;
      *)
        if test -n "$pick_args"
        then
          # This sed command makes sure that any special chars are quoted,
          # so the arg gets passed exactly to the server.
          args="$args "`echo "$arg" | sed -e 's,\([^a-zA-Z0-9_.-]\),\\\\\1,g'`
        fi
        ;;
    esac
  done
}


MY_PWD=`pwd`
# Check if we are starting this relative (for the binary release)
if test -d $MY_PWD/data/mysql -a -f ./share/mysql/english/errmsg.sys -a \
 -x ./bin/mysqld
then
  MY_BASEDIR_VERSION=$MY_PWD		# Where bin, share and data are
  ledir=$MY_BASEDIR_VERSION/bin		# Where mysqld is
  DATADIR=$MY_BASEDIR_VERSION/data
  if test -z "$defaults"
  then
    defaults="--defaults-extra-file=$MY_BASEDIR_VERSION/data/my.cnf"
  fi
# Check if this is a 'moved install directory'
elif test -f ./var/mysql/db.frm -a -f ./share/mysql/english/errmsg.sys -a \
 -x ./libexec/mysqld
then
  MY_BASEDIR_VERSION=$MY_PWD		# Where libexec, share and var are
  ledir=$MY_BASEDIR_VERSION/libexec	# Where mysqld is
  DATADIR=$MY_BASEDIR_VERSION/var
else
  MY_BASEDIR_VERSION=@prefix@
  DATADIR=@localstatedir@
  ledir=@libexecdir@
fi

user=@MYSQLD_USER@
niceness=0

# Use the mysqld-max binary by default if the user doesn't specify a binary
if test -x $ledir/mysqld-max
then
  MYSQLD=mysqld-max
else
  MYSQLD=mysqld
fi

# these rely on $DATADIR by default, so we'll set them later on
pid_file=
err_log=

# Get first arguments from the my.cnf file, groups [mysqld] and [mysqld_safe]
# and then merge with the command line arguments
if test -x ./bin/my_print_defaults
then
  print_defaults="./bin/my_print_defaults"
elif test -x @bindir@/my_print_defaults
then
  print_defaults="@bindir@/my_print_defaults"
elif test -x @bindir@/mysql_print_defaults
then
  print_defaults="@bindir@/mysql_print_defaults"
else
  print_defaults="my_print_defaults"
fi

args=
SET_USER=2
parse_arguments `$print_defaults --loose-verbose $defaults mysqld server`
if test $SET_USER -eq 2
then
  SET_USER=0
fi
parse_arguments `$print_defaults --loose-verbose $defaults mysqld_safe safe_mysqld`
parse_arguments PICK-ARGS-FROM-ARGV "$@"
safe_mysql_unix_port=${mysql_unix_port:-${MYSQL_UNIX_PORT:-@MYSQL_UNIX_ADDR@}}

if test ! -x $ledir/$MYSQLD
then
  echo "The file $ledir/$MYSQLD doesn't exist or is not executable"
  echo "Please do a cd to the mysql installation directory and restart"
  echo "this script from there as follows:"
  echo "./bin/mysqld_safe".
  echo "See http://dev.mysql.com/doc/mysql/en/mysqld_safe.html for more"
  echo "information"
  exit 1
fi

if test -z "$pid_file"
then
  pid_file=$DATADIR/`@HOSTNAME@`.pid
else
  case "$pid_file" in
    /* ) ;;
    * )  pid_file="$DATADIR/$pid_file" ;;
  esac
fi
test -z "$err_log"  && err_log=$DATADIR/`@HOSTNAME@`.err

if test -n "$mysql_unix_port"
then
  args="--socket=$mysql_unix_port $args"
fi
if test -n "$mysql_tcp_port"
then
  args="--port=$mysql_tcp_port $args"
fi

if test $niceness -eq 0
then
  NOHUP_NICENESS="nohup"
else
  NOHUP_NICENESS="nohup nice -$niceness"
fi

# Using nice with no args to get the niceness level is GNU-specific.
# This check could be extended for other operating systems (e.g.,
# BSD could use "nohup sh -c 'ps -o nice -p $$' | tail -1").
# But, it also seems that GNU nohup is the only one which messes
# with the priority, so this is okay.
if nohup nice > /dev/null 2>&1
then
    normal_niceness=`nice`
    nohup_niceness=`nohup nice`

    numeric_nice_values=1
    for val in $normal_niceness $nohup_niceness
    do
        case "$val" in
            -[0-9] | -[0-9][0-9] | -[0-9][0-9][0-9] | \
             [0-9] |  [0-9][0-9] |  [0-9][0-9][0-9] )
                ;;
            * )
                numeric_nice_values=0 ;;
        esac
    done

    if test $numeric_nice_values -eq 1
    then
        nice_value_diff=`expr $nohup_niceness - $normal_niceness`
        if test $? -eq 0 && test $nice_value_diff -gt 0 && \
            nice --$nice_value_diff echo testing > /dev/null 2>&1
        then
            # nohup increases the priority (bad), and we are permitted
            # to lower the priority with respect to the value the user
            # might have been given
            niceness=`expr $niceness - $nice_value_diff`
            NOHUP_NICENESS="nice -$niceness nohup"
        fi
    fi
else
    if nohup echo testing > /dev/null 2>&1
    then
        :
    else
        # nohup doesn't work on this system
        NOHUP_NICENESS=""
    fi
fi

USER_OPTION=""
if test -w / -o "$USER" = "root"
then
  if test "$user" != "root" -o $SET_USER = 1
  then
    USER_OPTION="--user=$user"
  fi
  # If we are root, change the err log to the right user.
  touch $err_log; chown $user $err_log
  if test -n "$open_files"
  then
    ulimit -n $open_files
    args="--open-files-limit=$open_files $args"
  fi
  if test -n "$core_file_size"
  then
    ulimit -c $core_file_size
  fi
fi

#
# If there exists an old pid file, check if the daemon is already running
# Note: The switches to 'ps' may depend on your operating system
if test -f $pid_file
then
  PID=`cat $pid_file`
  if @CHECK_PID@
  then
    if @FIND_PROC@
    then    # The pid contains a mysqld process
      echo "A mysqld process already exists"
      echo "A mysqld process already exists at " `date` >> $err_log
      exit 1
    fi
  fi
  rm -f $pid_file
  if test -f $pid_file
  then
    echo "Fatal error: Can't remove the pid file: $pid_file"
    echo "Fatal error: Can't remove the pid file: $pid_file at " `date` >> $err_log
    echo "Please remove it manually and start $0 again"
    echo "mysqld daemon not started"
    exit 1
  fi
fi

#
# Uncomment the following lines if you want all tables to be automatically
# checked and repaired during startup. You should add sensible key_buffer
# and sort_buffer values to my.cnf to improve check performance or require
# less disk space.
# Alternatively, you can start mysqld with the "myisam-recover" option. See
# the manual for details.
#
# echo "Checking tables in $DATADIR"
# $MY_BASEDIR_VERSION/bin/myisamchk --silent --force --fast --medium-check $DATADIR/*/*.MYI
# $MY_BASEDIR_VERSION/bin/isamchk --silent --force $DATADIR/*/*.ISM

echo "Starting $MYSQLD daemon with databases from $DATADIR"

# Does this work on all systems?
#if type ulimit | grep "shell builtin" > /dev/null
#then
#  ulimit -n 256 > /dev/null 2>&1		# Fix for BSD and FreeBSD systems
#fi

echo "`date +'%y%m%d %H:%M:%S  mysqld started'`" >> $err_log
while true
do
  rm -f $safe_mysql_unix_port $pid_file	# Some extra safety
  if test -z "$args"
  then
    $NOHUP_NICENESS $ledir/$MYSQLD $defaults --basedir=$MY_BASEDIR_VERSION --datadir=$DATADIR $USER_OPTION --pid-file=$pid_file @MYSQLD_DEFAULT_SWITCHES@ >> $err_log 2>&1
  else
    eval "$NOHUP_NICENESS $ledir/$MYSQLD $defaults --basedir=$MY_BASEDIR_VERSION --datadir=$DATADIR $USER_OPTION --pid-file=$pid_file @MYSQLD_DEFAULT_SWITCHES@ $args >> $err_log 2>&1"
  fi
  if test ! -f $pid_file		# This is removed if normal shutdown
  then
    break
  fi

  if test @IS_LINUX@ -a $KILL_MYSQLD -eq 1
  then
    # Test if one process was hanging.
    # This is only a fix for Linux (running as base 3 mysqld processes)
    # but should work for the rest of the servers.
    # The only thing is ps x => redhat 5 gives warnings when using ps -x.
    # kill -9 is used or the process won't react on the kill.
    numofproces=`ps xa | grep -v "grep" | grep -c $ledir/$MYSQLD`
    echo -e "\nNumber of processes running now: $numofproces" | tee -a $err_log
    I=1
    while test "$I" -le "$numofproces"
    do 
      PROC=`ps xa | grep $ledir/$MYSQLD | grep -v "grep" | sed -n '$p'` 
	for T in $PROC
	do
	  break
	done
	#    echo "TEST $I - $T **"
	if kill -9 $T
	then
	  echo "$MYSQLD process hanging, pid $T - killed" | tee -a $err_log
	else 
	  break
	fi
	I=`expr $I + 1`
    done
  fi
  echo "`date +'%y%m%d %H:%M:%S'`  mysqld restarted" | tee -a $err_log
done

echo "`date +'%y%m%d %H:%M:%S'`  mysqld ended" | tee -a $err_log
echo "" | tee -a $err_log
